### ISSUE TITLE
SCS abuses convert

### ISSUE THREAD

**Issue by odow** [jump-dev/SCS.jl#217]
Stuff like this
https://github.com/jump-dev/SCS.jl/blob/500e75d8f3a3467b8ae2814a49e41aa05c31439e/src/types.jl#L30-L52
makes me nervous.

We should thoroughly audit the use of cconvert in SCS.jl. We're passing a lot of `Ref{SCSData}` to `ccall`, when best practice is to implement `unsafe_convert` and `cconvert`. This should limit the need for `GC.@preserve` etc.

**Comment by kalmarek** [comment]
I made this change precisely to limit the need of `GC.@preserve` ;)
TL;DR: I don't think it's possible to create `SCSData` inside `ccall` (e.g. during `cconvert`) so I added `scsmatref` to ease the construction of `SCSData`. 


The point is that `SCSData.A` is a `Ptr{SCSMatrix}` and `SCSMatrix` contains pointers to `ManagedSCSMatrix` fields.
So You may either pass `ManagedSCSMatrix` (without `scsmatref`) to `ccall` and construct `SCSData` in `cconvert` (? i doubt this is even possible, since `SCSData` needs a ptr to `SCSSettings`), or you need to construct `SCSData` before `ccall` and then you need to `@GC.preserve` everything it contains pointers to. As the comment says, as long as `managed_matrix::ManagedSCSMatrix` is preserved, the `SCSData(..., managed_matrix, ...)` created from it will point to valid memory. 

Or you have a brand new idea how to solve this ;)

**Comment by odow** [comment]
I think I need to take a good look through the package, which I haven't done in any detail before. I understand why things are the way they are. But I don't know enough of the details yet to suggest an alternative.

**Comment by kalmarek** [comment]
please do so, I'm not very happy about the current solution as well. If you want we can have a chat next week
